### ISSUE TITLE
lint `vec.clone().into_iter().stuff().collect()` to suggest `vec.iter().stuff().cloned().collect()`

### ISSUE THREAD

**Issue by oli-obk** [rust-lang-nursery/rust-clippy#3302]
Found the following code in the wild

```rust
 let w: Vec<String> = vec_of_string
        .clone()
        .into_iter()
        .filter(|s| s.len() < 10)
        .collect();
```

Not sure what other things should be linted, but at least filter and filter_map after a clone+into_iter

**Comment by scottmcm** [comment]
I think this can be split into two parts:
- lint `.clone().into_iter()`, suggesting `.iter().cloned()` instead (to skip the container allocation)
- lint `.cloned().filter()`, suggesting `.filter().cloned()` instead (to skip filtered-out things)

**Comment by hcpl** [comment]
Throwing a wild idea into the mix: sort adaptors by their individual performance impact and then find a global optimum of the combination. I believe this will allow for a more generic lint. As for an implementation, a guided brute-force should do if the adaptor chain is short, say, < 10-15 elements; if not, it's too long anyway (is there a lint for that already?).

**Comment by GnomedDev** [comment]
I feel like that would be way too hard to implement and seems to have caused this idea to get lost for a couple years. A lint for `.clone().into_iter()` seems like quite an easy thing to implement.
